docs/reference/classes.md

```diff
@@ -309,3 +309,37 @@ inside that class.
 Even more specifically, if you declare a [companion object](object-declarations.html#companion-objects) inside your class,
 you'll be able to call its members with the same syntax as calling static methods in Java/C#, using only the class name
 as a qualifier.
+
+
+## Sealed Classes
+
+Sealed classes are used for representing restricted class hierarchies, when a value can have one of the types from a
+limited set, but cannot have any other type. They are, in a sense, an extension of enum classes: the set of values
+for an enum type is also restricted, but each enum constant exists only as a single instance, whereas a subclass
+of a sealed class can have multiple instances which can contain state.
+
+To declare a sealed class, you put the `sealed` modifier before the name of the class. A sealed class can have
+subclasses, but all of them must be nested inside the declaration of the sealed class itself.
+
+``` kotlin
+sealed class Expr {
+    class Const(val number: Double) : Expr()
+    class Sum(e1: Expr, e2: Expr) : Expr()
+    object NotANumber : Expr()
+}
+```
+
+Note that multiple levels of inheritance are also allowed, as long as the entire hierarchy is nested within the
+enclosing sealed class.
```

Comment:
I doubt this is correct. Doesn't `sealed` only restrict direct subclasses, and has no idea about indirect ones? // cc @mglukhikh